### PR TITLE
Use checksums to disambiguate snapshots in >1 pack

### PR DESCRIPTION
**Pull request by veselink1**
This PR takes care of non-unique snapshots/snapshots available in multiple packs by comparing content checksums. If the snapshots are determined to be non-identical, `AmbiguousSnapshotMatch` is returned, as before.

Potentially relevant to #74, depending on how we want to implement `elfshaker list`.

### CODE REVIEW

File: src/packidx.rs [hunk 1]
```diff
@@ -439,6 +442,32 @@ impl PackIndex {
...
+}
+
+#[cfg(unix)]
+fn os_str_as_bytes(os_str: &OsStr) -> Option<&[u8]> {
+    Some(std::os::unix::ffi::OsStrExt::as_bytes(os_str))
+}
+
+#[cfg(not(unix))]
+fn os_str_as_bytes(os_str: &OsStr) -> Option<&[u8]> {
+    None()
```
Comment (pwaller) [inline src/packidx.rs:470]: I think this warrants a comment at least. I read that the effect of it is that on windows, two snapshots with the same name, with an identical set of checksums in it, but different path names, would be declared equivalent where they are not.

My preference would be that we don't use OsStr in our internal paths, but clearly that's a harder fix. So I'm willing to live with this minor gotcha as-is for now.

File: src/packidx.rs [hunk 2]
```diff
@@ -439,6 +442,32 @@ impl PackIndex {
         }
         Ok(None)
     }
+    /// Computes the checksum of the contents of the snapshot.
+    pub fn compute_snapshot_checksum(&self, snapshot: &str) -> Option<ObjectChecksum> {
+        // Get the FIleHandles
```
Comment (pwaller) [inline src/packidx.rs:447]: Think this comment can be dropped without loss.

File: src/repo/repository.rs [hunk 3]
```diff
@@ -912,6 +912,35 @@ impl Repository {
...
+            .map(|pack| self.load_index(pack).map(|packidx| {
+                packidx.compute_snapshot_checksum(snapshot)
+            })
+            .expect("failed to resolve snapshot"));
+
+        let first = snapshot_checksums.next().expect("At least 1 pack expected");
+        Ok(snapshot_checksums.all(|checksum| checksum == first))
+    }
+
+    /// Tries to disambiguate the snapshot between the packs providing it by
```
Comment (pwaller) [inline src/repo/repository.rs:928]: Suggestion for docstring:
> `disambiguate_snapshot` takes a set of Packs which contain the given snapshot name. If the snapshot in each pack is identical according to content checksum, return an arbitrary pack, preferring a loose one if available.

**Review by pwaller — APPROVED** [review]
Looks good, some minor optional suggestions.

File: src/packidx.rs (hunk 1 above)
Comment (veselink1) [reply src/packidx.rs]: I "fixed" this by using to_string_lossy, which may allocate but is at least more correct.

**Comment by veselink1** [comment]
Thanks for the review and suggestions.

**Pull request merged**